### DISCUSSION TITLE
Add OCIO presets.

### DISCUSSION THREAD

**Discussion by AltaArts** [ggarra13/mrv2#256] [Ideas]
First, since it is my first post here let me say MRV2 looks like a great replacement viewer for us.  We had been using DJV2, but needed OCIO2 support.  RV is clunky and a pain, and Cinesync is good, but it keeps clipping >1 images even when using OCIO.  Plus they are not being super reactive in suggestions.

So saying all that, it seems like MRV2 meets all of those!  And for that I thank you.

Now for the first suggestion:  OCIO presets in the top tool bar.  We get material with several different "colorspaces" (Rec709 24, Rec709 Linear, ACEScg, ACEScct, various camera log, etc).  We are using a semi-custom OCIO2 which is working very well with MRV2, but it does take some time to pick the "IDT", "ODT", "Look", and sometimes LUT when opening a media file.  It would be great if we could save user presets that have those all, including a LUT from a local dir.   This is something Cinesync does have, even though it is stuck under a menu.

I would think it would live next to the other buttons for colorspace in the top bar, and preferably have a keyboard shortcut to bring up the menu - that way we can have the top bar hidden as well.

Again, thanks so much.

JRB.

**Comment by AltaArts** [comment]
Just wanted to bump this to ask if it is on the roadmap?

Thanks much.

**Reply by ggarra13** [reply]
> Just wanted to bump this to ask if it is on the roadmap?
> 
> Thanks much.

I did an attempt at implementing it and it ended up being a mess so I threw it away.  So for now, I am going to say no.

There are several issues:

1) Whether the current selected Preset should get saved and start as active.  I am inclining for saying strongly "no" to this, as it makes things really hard (that's mainly what I implemented an why it ended up being a mess).
2) Screen space.  Adding it to the top bar presents a problem when mrv2 starts as I don't have screen space, except when the window maximizes to 4K or similar.  I am afraid I would have to place it in either Preferences or the Color Panel.  Both options  make it hard (Preferences for updating and Color Panel to add a button to create/delete presets).

At this point in time, if you need it asap my suggestion is that you implement it with a Python plug-in.  With a python plug-in, you can add a menu entry for your common settings and change the OCIO settings from there.  No shortcuts thou.

**Comment by ggarra13** [comment]
I did a second attempt at implementing it and this time it came out good.  It will be there when I release v1.2.1.

**Comment by AltaArts** [comment]
Ok, nice.  Yesterday I did a quick plugin for what I was thinking.  I think it is working pretty well, and have some things to improve.  Is this the sorta thing you were thinking?  In a perfect world, I would have the preset "menu" right next to the other OCIO buttons, and have a hot key to open the presets menu.

https://github.com/AltaArts/OCIO_Presets--MRV2-Plugin

**Comment by ggarra13** [comment]
That's very cool.  And you managed to use pyfltk too.  Anyway, here's the current workflow in my v1.2.1 beta:


https://github.com/ggarra13/mrv2/assets/6072843/d198287d-87f8-475c-a6d9-5932cdd85aed

**Comment by AltaArts** [comment]
That is exactly what I was thinking.  That fits right into the main core nicely.

Question:  can there be a button and/or hotkey to call the saved presets?  We tend to have to view images in different transforms, so we can't really use the defaults (8, 16, 32, etc) unless I am missing something.  For example one exr in half-float Rec709Linear, and another in half-float ACEScg.  Going into the menus to change the preset for each is a bit of a pain.

I was thinking it could be just a small button to the left of the OCIO input button. But if the screen space is the issue (I do not know fltk really), then maybe the button and hotkey just opens a popup to choose - that is where I was going to head.

Also, I was going to add a check of the filename to be able to set defaults.  Meaning if the filename has "ACEScg" in the name, it would automatically set the preset to that.

I would be very happy to give your version a test on our machines/config.

**Reply by ggarra13** [reply]
> I would be very happy to give your version a test on our machines/config.

You can try the beta version that is in sourceforge at:

https://sourceforge.net/projects/mrv2/files/beta/ 

> Question: can there be a button and/or hotkey to call the saved presets? We tend to have to view images in different transforms, so we can't really use the defaults (8, 16, 32, etc) unless I am missing something.

Each preset currently also saves the 8, 16, 32, etc. defaults for it, besides the OCIO and LUT settings.

> Also, I was going to add a check of the filename to be able to set defaults. Meaning if the filename has "ACEScg" in the name, it would automatically set the preset to that.

Currently, all presets are saved in a single .json file.

**Comment by AltaArts** [comment]
Fab.  I will get it here in a bit.  Just finishing some stuff. 

Thanks so much for working on it!

**Comment by AltaArts** [comment]
@ggarra13 :

Downloaded the beta and tried it out.  I think it is working very well!  Thanks for the hot key, and with that the Preset Panel pops up super fast.  I quickly built some presets and it seems to work perfectly with our OCIO, and even the luts.  Can I ask where is the config .json saved?  I clicked around but couldn't find it.  It might be great to be able to share it amongst our machines instead of rebuilding.

The one thing I did see (I know it is beta, but wanted to let you know just in case), is the UI styling of the Preset Panel appears to be messed up.

![Preset Panel](https://github.com/ggarra13/mrv2/assets/86539171/4ddf6234-1df8-480f-81f0-0a2884912cbb)


And I am sure you have done color tests, but here are some.  You can see that MRV is probably the best, with RV coming a close second (if not the same).  As you can see CineSync has an issue with values > 1, and DJV2 looks different since I cannot use our OCIO2 config with it (it is still using ACES 1.2 IIRC). 

Basically, I had a shot laying around that had some good range in log that I also already had an exr seq in ACEScg.  I wanted to just do a screen capture off my grading screen to show what is actually being displayed apples to apples (all using the same OCIO transforms, except DJV2 as stated above).  The captures then went into Resolve with a difference matte.

MRV2:

![MRV2_ACEScg_EXR16](https://github.com/ggarra13/mrv2/assets/86539171/81cdbb85-d332-4008-bdca-ece9a2214b33)    ![MRV2_diff](https://github.com/ggarra13/mrv2/assets/86539171/efa13eae-488b-445e-b4cb-add3e16b35e8)

RV:

![RV_ACEScg_EXR16](https://github.com/ggarra13/mrv2/assets/86539171/c3c8b147-8d3a-43bb-8489-0b54335b458e)    ![RV_diff](https://github.com/ggarra13/mrv2/assets/86539171/56f89825-75cf-4e14-b896-eb67015c74b2)

CineSync:

![CineSync_ACEScg_EXR16](https://github.com/ggarra13/mrv2/assets/86539171/3b574678-6a9d-4c7e-9b85-b3ca3cd6495b)    ![CineSync_diff](https://github.com/ggarra13/mrv2/assets/86539171/f4621104-c299-4d94-9249-4dd81374f09c)

DJV2:

![DJV2_ACEScg_EXR16](https://github.com/ggarra13/mrv2/assets/86539171/760abb40-9140-4d73-b04d-888c40bf05f8)    ![DJV2_diff](https://github.com/ggarra13/mrv2/assets/86539171/10b8cb41-5a33-4bfb-8d1d-6ab5eb555b8e)

**Reply by ggarra13** [reply]
> Can I ask where is the config .json saved? I clicked around but couldn't find it. It might be great to be able to share it amongst our machines instead of rebuilding.

It is saved in your home dir (.filmaura subdir).

It is loaded, however, in this order:

```
   %HOME%/.filmaura/mrv2.ocio
   %USERPROFILE%/.filmaura/mrv2.ocio
   %HOMEDRIVE%/%HOMEPATH%/%USERNAME%/.filmaura/mrv2.ocio
```
   
You can open the Log Panel to see exactly where it is loaded from.  

NOTE:  I'll probably will be renaming the "mrv2.ocio" file to "mrv2.ocio.json" before the release.
